src/xunit.runner.reporters/VstsClient.cs

```diff
@@ -0,0 +1,310 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Xunit.Runner.Reporters
+{
+    public class VstsClient
+    {
+        readonly IRunnerLogger logger;
+        readonly string baseUri;
+        readonly int buildId;
+        static readonly HttpMethod PatchHttpMethod = new HttpMethod("PATCH");
+        const string UNIQUEIDKEY = "UNIQUEIDKEY";
+
+        public VstsClient(IRunnerLogger logger, string baseUri, string accessToken, int buildId)
+        {
+            this.logger = logger;
+            this.baseUri = baseUri;
+            this.buildId = buildId;
+            
+            client = new HttpClient();
+            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+
+            workTask = Task.Run(RunLoop);
+        }
+
+        readonly HttpClient client;
+        readonly MediaTypeWithQualityHeaderValue jsonMediaType = new MediaTypeWithQualityHeaderValue("application/json");
+
+        readonly Task workTask;
+        volatile bool previousErrors;
+
+        readonly ManualResetEventSlim finished = new ManualResetEventSlim(false);
+        readonly ManualResetEventSlim workEvent = new ManualResetEventSlim(false);
+        volatile bool shouldExit;
+
+        ConcurrentQueue<IDictionary<string, object>> addQueue = new ConcurrentQueue<IDictionary<string, object>>();
+        ConcurrentQueue<IDictionary<string, object>> updateQueue = new ConcurrentQueue<IDictionary<string, object>>();
+
+        ConcurrentDictionary<string, int> unqiueIdToTestIdMap = new ConcurrentDictionary<string, int>();
+
+        public void WaitOne(CancellationToken cancellationToken)
+        {
+            // Free up to process any remaining work
+            shouldExit = true;
+            workEvent.Set();
+
+            finished.Wait(cancellationToken);
+            finished.Dispose();
+        }
+
+        async Task RunLoop()
+        {
+            int? runId = null;
+            try
+            {
+                runId = await CreateTestRun();
+
+                while (!shouldExit || !addQueue.IsEmpty || !updateQueue.IsEmpty)
+                {
+                    workEvent.Wait();   // Wait for work
+                    workEvent.Reset();  // Reset first to ensure any subsequent modification sets
```

Comment:
This is a race condition. If you want this behavior, switch from `ManualResetEventSlim` to `AutoResetEventSlim`.

Reply:
We use the same pattern in `AppVeyorClient`, I assume it needs updating there too?